### PR TITLE
CRS-1766 Draft

### PR DESCRIPTION
**Pull request by jbowdidge-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/CalculableSentence.kt [hunk 1]
```diff
@@ -188,6 +188,9 @@ interface CalculableSentence {
 
   fun calculateErsedFromHalfway(): Boolean
   fun calculateErsedFromTwoThirds(): Boolean
+  fun calculateMixed(): Boolean {
+    return false
+  }
```
Comment (yasinmustafa) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/CalculableSentence.kt:193]: minor but could replace with just one line
`fun calculateMixed(): Boolean = false`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt [hunk 2]
```diff
@@ -266,29 +310,90 @@ data class SentenceCalculation(
...
       }
+      if (sentence.calculateMixed()) {
+        return calculateErsedMixed()
+      }
     }
     return null
   }
 
+  private fun calculateErsedMixed(): ReleaseDateCalculationBreakdown? {
+    val days = if (sentence is ConsecutiveSentence) {
```
Comment (ldlharper) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt:321]: Is looks like this is unused?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ErsedCalculator.kt [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.Offender
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.ReleaseDateCalculationBreakdown
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.SentenceCalculation
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.util.isAfterOrEqualTo
+import java.time.temporal.ChronoUnit
+import kotlin.math.ceil
+import kotlin.math.max
+
+@Service
+class ErsedCalculator(val ersedConfiguration: ErsedConfiguration) {
```
Comment (ldlharper) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ErsedCalculator.kt:19]: I can't see where this is used? Its mostly talking about HDCED too.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt [hunk 4]
```diff
@@ -266,29 +310,90 @@ data class SentenceCalculation(
...
+        sentence.sentencedAt,
+      )
     } else {
       val custodialDuration = sentence.custodialDuration()
       custodialDuration.getLengthInDays(sentence.sentencedAt)
     }
     return if (days >= RELEASE_AT_TWO_THIRDS_ERSED_DAYS) {
       val release = extendedDeterminateParoleEligibilityDate ?: releaseDate
-      val ersed = release.minusYears(1)
+      val ersed = release.minusDays(MAX_ERSED_PERIOD_DAYS.toLong())
```
Comment (yasinmustafa) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt:382]: nice work moving these hard coded numbers to constants - makes it easier to read

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/CalculableSentence.kt (hunk 1 above)
Comment (ldlharper) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/CalculableSentence.kt]: Also should probably be named related to ERSED.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ErsedCalculator.kt (hunk 3 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ErsedCalculator.kt:19]: Yup - we'll remove this one as this was a desired change but we won't do it now due to time constraints.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt (hunk 4 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt:373]: 🙌

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt (hunk 2 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt]: yep - removing.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/CalculableSentence.kt (hunk 1 above)
Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/CalculableSentence.kt]: Nice comments - doing now.

**Comment by yasinmustafa** [comment]
LGTM - some minor comments

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ErsedCalculator.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (srowlands-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/ErsedCalculator.kt]: It isn't, started the abstraction but then focussed on calc and testing.
@jbowdidge-moj has removed for now, we can make it nicer later

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt [hunk 5]
```diff
@@ -266,29 +310,90 @@ data class SentenceCalculation(
...
+    val days = if (sentence is ConsecutiveSentence) {
+      ConsecutiveSentenceAggregator((sentence as ConsecutiveSentence).orderedSentences.map { it.custodialDuration() }).calculateDays(
+        sentence.sentencedAt,
+      )
+    } else {
+      val custodialDuration = sentence.custodialDuration()
+      custodialDuration.getLengthInDays(sentence.sentencedAt)
+    }
+
+    val effectiveRelease = extendedDeterminateParoleEligibilityDate ?: adjustedDeterminateReleaseDate
```
Comment (ldlharper) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt:321]: It looks like this is duplicated in a few places. Where the ersed is calculated by the max period before release. Might be good to reduce the duplication in a new function.

```
val release = extendedDeterminateParoleEligibilityDate ?: releaseDate
      val ersed = release.minusDays(MAX_ERSED_PERIOD_DAYS.toLong())
      ReleaseDateCalculationBreakdown(
        rules = setOf(CalculationRule.ERSED_MAX_PERIOD),
        releaseDate = ersed,
        unadjustedDate = unadjustedExtendedDeterminateParoleEligibilityDate ?: unadjustedDeterminateReleaseDate,
        adjustedDays = ChronoUnit.DAYS.between(
          unadjustedDeterminateReleaseDate,
          adjustedDeterminateReleaseDate,
        ).toInt(),
        rulesWithExtraAdjustments = mapOf(
          CalculationRule.ERSED_MAX_PERIOD to AdjustmentDuration(
            -MAX_ERSED_PERIOD_DAYS,
            ChronoUnit.DAYS,
          ),
        ),
      )
      ```

Comment (jbowdidge-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/model/SentenceCalculation.kt:321]: Ordinarily - we would refactor this. I think its more important we get this into dev at the moment given the timeline.

**Review by ldlharper — APPROVED** [review]
LGTM

**Pull request merged**